Changing Fields to Properties and fixing singleton instances

BraedonWooding opened the pull request:
- This is going to be a pretty big PR (though I may split it up into multiple).  
- All it does is start to change all the fields into properties maybe 2/3 of the project is wrong in this aspect.  
- So far this completes all the controllers and DevConsole (along with maybe a few others).  
- This also removes the singleton instance for SceneController since everything was already static... and makes that class a static.
- I will work on 'removing' more singletons (through often making the private and having guard statements in public static functions), the power of having private singletons is mainly around controlling what people can do with your class.  Since (as found with build mode controller) if you have to access fields and edit them to perform an action either you're missing a function or you have a lot of hard coding.
- A good example of the benefits of property changing is build mode controller, previously to do a 'build' via code (something that a dev console command did) you had to set the build mode and type yourself and tile (if you building a tile) then call 'build' then unset them.  This is messy and doesn't really fit in, so I just made a generic build mode command to set the build mode, the others still exist but most likely I'll deprecate them out.

koosemose commented:
I would definitely advise you to try to split things up (or at least further things. Large sweeping changes like this can be a pain to overview and keep track of what's happening, and you also have two separate "focuses" here, removing singletons, and swapping public fields for public properties, I would prefer to be able to review the removing of singletons on it's own rather than mixed with something else, I'm not entirely sold that it is necessary, and that it's not just a matter of a differing style. Also, remember that it's only publics that need to be made properties. 

When you're making a change such as the one you reference in BMC, it would be advisable to do in a separate PR, since it sounds like it's changing functionality or at least how things interact, beyond just field or property

As one final thing to overview, when you're making large changes like this, you will almost certainly have to make changes to the "mod" files, both lua and C#, as those aren't in the assembly monodevelop (and I assume visual studio) won't realise they're using old versions of things (or non-existent statics).

BraedonWooding commented:
In cases where it's not needed like in scene manager, it just adds confusion.  In other cases where it's not 'better' objectively I'll avoid from now on.  And yes I've only made publics so far (converted some publics to privates too that shouldn't be exposed).  

It doesn't change functionality just adds a more generic variant along with the specific ones (like there is a 'ChangeBuildModeToFurniture' and so on and I just added another one just to handle them all, no other code changes), the others should 'probably' be removed as I said but I did highlight deprecate to try to hint (probably badly) at the idea that it is out of the scope of this PR and should be done in a separate one.

I've changed the C# ones to be updated (which are easy since they error at runtime) and I've looked over the LUA but I'll have to at some point do a more intensive LUA check.

koosemose commented on Assets/Scripts/Controllers/Events/TradeController.cs:
```diff
@@ -54,18 +54,12 @@ public void CallTradeShipTest(Furniture landingPad)
-        controller.AnimationFlying = prototype.AnimationFlying.Clone();
-        controller.AnimationIdle = prototype.AnimationIdle.Clone();
-        controller.Renderer = spriteRenderer;
+        controller.Init(leavingCoords: new Vector3(100, 50, 0), landingCoords: new Vector3(landingPad.Tile.X + 1, landingPad.Tile.Y + 1, 0), speed: 5f, trader: trader, animationIdle: prototype.AnimationIdle.Clone(), animationFlying: prototype.AnimationFlying.Clone(), renderer: spriteRenderer);
```
Put a `TODO:` here, it looks like too much work is being put on the caller of the tradeship, and I suspect more of the work could be done in init or elsewhere in the controller, so that calling a tradeship is simpler on the calling side. Something to the effect of `// TODO: Look into passing more of the work of calling a tradeship off to the controller`. I think it's outside the scope of this PR, but I'd prefer one of to not have to waste time digging back through the code to find it again, or just forget about it completely.

BraedonWooding replied:
Yep, I agree.  I'll place one in.  If only constructors worked on Monobehaviours... I just made it as 'simple' since I didn't want to branch out too much (out of scope).

koosemose commented on Assets/Scripts/UI/InGameUI/MenuTop/CurrencyDisplay.cs:
```diff
@@ -13,7 +13,8 @@
 {
-    public Text text;
+    [SerializeField]
+    private Text text;
```
Does this even need serialized, I realize adding that keeps closer to the previous behavior, but since we do so little work in editor, I'm not sure it even needs serialized (serialization is the editor is the only benefit I can think of for this).

BraedonWooding replied:
I believe it is set in the editor, I will double check that.

BraedonWooding replied:
And that is what it was done for.

koosemose commented:
Ok I've done a full visual review of the code, it looks good, aside from the question relating to serialization of a field. If any of these classes are adjacent to prefabs (either used in or store a reference to) make sure to double check that related functions still run in play mode, since sometimes altering scripts adjacent to prefabs can break editor based references.

Also I'd request calling this pass complete for this PR, mostly for selfish reasons, in that this is near the maximum length I feel I can reliably visually review before losing focus, and because I'd rather not end up going over code I've already looked over.

BraedonWooding commented:
Yep already checked references and scripts.  And yes I agree.  I'll pull it in and do documentation in another PR (at one point).

BraedonWooding commented:
Okay so it was needed, however taking in what you said which I completely agreed with I removed the `[SerializeField]` and rather just add a `[RequireComponent(typeof(Text))]` to the class along with an awake call that just says `text = GetComponent<Text>();`, which emulates old functionality without requiring editor support (and well it should always have a text component so that 'helps').  No other lines changed.  If the check passes I'll pull it in.

koosemose commented:
We need to be on the watch for bugs caused by the mass changing of public fields to properties, as at least some of those held references via Unity, that are no longer being set (current case being CircleCursorPrefab no longer holding a reference to the appropriate prefab.)
